### PR TITLE
Use test filter files to exclude upstream tests from running

### PR DESCRIPTION
**Pull request by mariospr**
This PR drops all the many chromium_src overriden files to empty browser
and unit tests and exclude them and moves into using Chromium's *.filter
files instead, which is an easier and more convenient way to maintain
exclusions.

But for those .filter files to be applied, this PR also adapts the
npm run test command so that it will correctly pass the right paths
to GTest when running upstream unit and/or browser tests, pointing to
such .filter files, if they exist. Note that it is possible to specify
general exclusions to be applied to all platforms and architectures as
well as platform-specific and arch-specific ones, and npm run test
will automatically select at runtime the ones that apply to the current
execution from those available in //brave/test/filters.

Last, this PR also includes up-to-date exclusions for all platforms, as
well as a few specific ones for Linux, Windows and MacOS, meaning that
we are now successfully running ~12K unit tests and ~8000K browser tests
from upstream when the CI/run-upstream-tests GitHub label is added. Only
exception is MacOS, where we still run upstream's unit tests but where 
we had to entirely disable the execution of all browser tests as it was
taking way too long (more than 6h) and being extremely flaky.

Resolves https://github.com/brave/brave-browser/issues/20456

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

N/A

### CODE REVIEW

**Comment by mariospr** [comment]
@bridiver @iefremov @mkarolin This is **finally ready for review**, now that we have managed to get a few runs successfully passing on CI after finding the right exclusions + disabling the MacOS CI bot for browser tests (thanks @wknapik for his unvaluable help with all this!)

PTAL when you have a moment, thanks!

**Comment by mariospr** [comment]
@bridiver @iefremov @mkarolin Ping again, thanks!

File: browser/brave_ads/ads_tab_helper.cc
Comment (mkarolin) [inline browser/brave_ads/ads_tab_helper.cc:53]: Not sure about this one: we only seem to use `ISOLATED_WORLD_ID_BRAVE_INTERNAL` with AdBlock (https://github.com/brave/brave-core/blob/master/renderer/brave_content_renderer_client.cc#L66), for ads we just use `content::ISOLATED_WORLD_ID_GLOBAL` (https://github.com/brave/brave-core/blob/master/renderer/brave_content_renderer_client.cc#L82), but I might be looking at it all wrong :-/

Comment (mariospr) [reply browser/brave_ads/ads_tab_helper.cc:53]: @mkarolin You could be perfectly right, I'm not sure about this bit either, just that I need to set the ID or otherwise all browser tests (or nearly all) will crash, but perhaps `ISOLATED_WORLD_ID_BRAVE_INTERNAL` is not the right one.

Cc @bridiver in case he knows the right answer here.

**Comment by wknapik** [comment]
ping

File: browser/brave_ads/ads_tab_helper.cc
Comment (bridiver) [reply browser/brave_ads/ads_tab_helper.cc:53]: dom distiller would normally be `ISOLATED_WORLD_ID_CHROME_INTERNAL`, but you shouldn't have to set this here, it's handled by `dom_distiller::RegisterViewerSource(this);` in ProfileImpl

Comment (bridiver) [reply browser/brave_ads/ads_tab_helper.cc:53]: it's not clear to me why this change would be needed for filtering upstream tests from running. If it's failing in a test, it's probably because the test profile doesn't call `dom_distiller::RegisterViewerSource(this);` so I would look at the dom distiller tests in upstream

Comment (mariospr) [reply browser/brave_ads/ads_tab_helper.cc:53]: @bridiver The problem is not related to just a subset of upstream tests, but to **all of them**, since they will all crash with a backtrace like the one in the commit message as soon as `AdsTabHelper::RunIsolatedJavaScript()` is run, where `dom_distiller::RunIsolatedJavaScript()` is called from:
```c++
void AdsTabHelper::RunIsolatedJavaScript(
    content::RenderFrameHost* render_frame_host) {
  DCHECK(render_frame_host);

  dom_distiller::RunIsolatedJavaScript(
      render_frame_host, "new XMLSerializer().serializeToString(document)",
      base::BindOnce(&AdsTabHelper::OnJavaScriptHtmlResult,
                     weak_factory_.GetWeakPtr()));

  dom_distiller::RunIsolatedJavaScript(
      render_frame_host, "document?.body?.innerText",
      base::BindOnce(&AdsTabHelper::OnJavaScriptTextResult,
                     weak_factory_.GetWeakPtr()));
}
```

See for instance `RegionCaptureBrowserTest.MaxCropIdsInTopLevelDocument` (but every other tests crashes in a similar way):
```
[ RUN      ] RegionCaptureBrowserTest.MaxCropIdsInTopLevelDocument

[2311890:2311890:0216/120832.162529:FATAL:distiller_javascript_utils.cc(41)] Check failed: distiller_javascript_world_id != invalid_world_id. 
#0 0x7f16475ee239 base::debug::CollectStackTrace()
#1 0x7f16474e46d3 base::debug::StackTrace::StackTrace()
#2 0x7f16475070b4 logging::LogMessage::~LogMessage()
#3 0x7f1647507b2e base::internal::LoggerWithAllowedAllocations::~LoggerWithAllowedAllocations()
#4 0x56308082daf0 dom_distiller::RunIsolatedJavaScript()
#5 0x563081a6fb28 brave_ads::AdsTabHelper::RunIsolatedJavaScript()
#6 0x7f164117781d content::WebContentsImpl::WebContentsObserverList::NotifyObservers<>()
#7 0x7f16411977f9 content::WebContentsImpl::DocumentOnLoadCompleted()
#8 0x7f163ea80b31 blink::mojom::LocalFrameHostStubDispatch::Accept()
#9 0x7f1642012cbd mojo::InterfaceEndpointClient::HandleValidatedMessage()
#10 0x7f164201913d mojo::MessageDispatcher::Accept()
#11 0x7f1642014a3d mojo::InterfaceEndpointClient::HandleIncomingMessage()
#12 0x7f1641a605fd IPC::(anonymous namespace)::ChannelAssociatedGroupController::AcceptOnEndpointThread()
#13 0x7f1641a5cd5c base::internal::Invoker<>::RunOnce()
#14 0x7f164757c8d9 base::TaskAnnotator::RunTaskImpl()
#15 0x7f16475a28ab base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl()
#16 0x7f16475a205b base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
#17 0x7f16475a3002 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
#18 0x7f16475160ef base::MessagePumpGlib::Run()
#19 0x7f16475a3560 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run()
#20 0x7f16475506b7 base::RunLoop::Run()
#21 0x7f164104a79e content::BrowserTaskExecutor::RunAllPendingTasksOnThreadForTesting()
#22 0x563080adf849 content::RunAllPendingInMessageLoop()
#23 0x563080adf808 content::RunAllPendingInMessageLoop()
#24 0x5630801810fc InProcessBrowserTest::PreRunTestOnMainThread()
#25 0x563080a69f1f content::BrowserTestBase::ProxyRunTestOnMainThreadLoop()
#26 0x7f1640b0e1c0 content::BrowserMainLoop::RunMainMessageLoop()
#27 0x7f1640b10312 content::BrowserMainRunnerImpl::Run()
#28 0x7f1640b0bb4b content::BrowserMain()
#29 0x7f1641626725 content::RunBrowserProcessMain()
#30 0x7f1641627fa3 content::ContentMainRunnerImpl::RunBrowser()
#31 0x7f164162792a content::ContentMainRunnerImpl::Run()
#32 0x7f1641624896 content::RunContentProcess()
#33 0x7f16416252a7 content::ContentMain()
#34 0x563080a69512 content::BrowserTestBase::SetUp()
#35 0x56308017fc4e InProcessBrowserTest::SetUp()
#36 0x56307f39ea10 testing::Test::Run()
#37 0x56307f39ffe3 testing::TestInfo::Run()
#38 0x56307f3a0db7 testing::TestSuite::Run()
#39 0x56307f3b51b7 testing::internal::UnitTestImpl::RunAllTests()
#40 0x56307f3b4afd testing::UnitTest::Run()
#41 0x5630801a1fe9 base::TestSuite::Run()
#42 0x563080148d52 ChromeTestSuiteRunner::RunTestSuite()
#43 0x563080adbf7a content::LaunchTests()
#44 0x56308014907c LaunchChromeTests()
#45 0x563080146575 main
#46 0x7f1634ef0fd0 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2dfcf)
#47 0x7f1634ef107d __libc_start_main
#48 0x56307d0d0cea _start
Task trace:
#0 0x7f1641a5bd0e IPC::(anonymous namespace)::ChannelAssociatedGroupController::Accept()
#1 0x7f1646d0594c mojo::SimpleWatcher::Context::Notify()
Crash keys:
  "ui_scheduler_async_stack" = "0x7F1641A5BD0E 0x7F1646D0594C"
  "io_scheduler_async_stack" = "0x7F1646D0594C 0x0"
```

It seems to me that the problem is that `brave_ads::AdsTabHelper::RunIsolatedJavaScript()` will get called no matter what via the WebContentsObserver machinery, but without having set the JS World ID first, which is why I added this but maybe that's not the right place after all.

I'll look into whether there's a better place to add that initialization then (and will use `ISOLATED_WORLD_ID_CHROME_INTERNAL` if needed), let's see... but the important thing is that this (or a equivalent) change is not needed for specific tests, but for all of them :innocent:

Comment (mariospr) [reply browser/brave_ads/ads_tab_helper.cc:53]: So I think I found the actual root problem, and it's indeed that there were some initialization-related steps missing when running the upstream tests:

Basically, when we run brave_browser_tests we rely on the launcher at `//brave//test/base/browser_tests_main.cc`, which associates an instance of `BraveTestLauncherDelegate` to the `ChromeTestSuiteRunner`. That delegate will then instantiate a `BraveMainDelegate` when `ChromeLauncherDelegate::CreateContentMainDelegate()` gets called, meaning that `BraveMainDelegate::BasicStartupComplete()` will get called later on during the initialization process, which makes sure that several features get enabled by default in Brave, including the DOM Distiller.

Then, when `ProfileImpl::DoFinalInit()` calls `dom_distiller::RegisterViewerSource()`, the JS World ID will be set from there because the feature is enabled and therefore that function won't early return:
```c++
void RegisterViewerSource(Profile* profile) {
  if (!dom_distiller::IsDomDistillerEnabled())
    return;

  LazyDomDistillerService* lazy_service =
      LazyDomDistillerService::Create(profile);

  // Set the JavaScript world ID.
  if (!DistillerJavaScriptWorldIdIsSet())
    SetDistillerJavaScriptWorldId(ISOLATED_WORLD_ID_CHROME_INTERNAL);

  content::URLDataSource::Add(
      profile, std::make_unique<DomDistillerViewerSource>(lazy_service,
                                                          kDomDistillerScheme));
}
```

**However**, when running upstream tests, the launcher in use is `//chrome/test/base/browser_tests_main.cc`, which instantiates a `ChromeTestLauncherDelegate` instead... which then means that a `ChromeMainDelegate` will be instantiated later on instead of our `BraveMainDelegate`. Due to this, we won't get `BraveMainDelegate::BasicStartupComplete()` called but `ChromeMainDelegate::BasicStartupComplete()`instead, meaning that we wont' get all those features we want in Brave enabled... including the DOM distiller. Finally, when `ProfileImpl::DoFinalInit()` calls `dom_distiller::RegisterViewerSource()`, the JS World ID will **NOT** be set from there because the feature is disabled and the function will early return, taking us to the point where we are now.

Therefore I think the proper solution is to make sure that at least whatever we have now in `BraveMainDelegate::BasicStartupComplete()` gets also used when running upstrem browser tests. That way we would get the DOM Distiller feature enabled which would fix this particular problem. Additionally, it would also make sure that the upstream browser tests are run with the same features configuration we have for Brave, which seems the correct thing to do too.

Comment (mariospr) [reply browser/brave_ads/ads_tab_helper.cc]: @mkarolin @bridiver I reverted the original patch that manually set the JS World ID and implemented another solution based the findings from my previous comment :point_up: in this commit: https://github.com/brave/brave-core/pull/11814/commits/72b44e806be51463abddf75e3f07e57101c59f64

PTAL when you have a moment, thanks!

**Comment by mariospr** [comment]
Rebased on master (now Chromium 99)

File: browser/brave_ads/ads_tab_helper.cc
Comment (mariospr) [reply browser/brave_ads/ads_tab_helper.cc]: @bridiver PR is ready for review again after addressing your comments + rebasing on top of Chromium 99

**Comment by mariospr** [comment]
@iefremov @bridiver This is ready for another review after addressing Brian's comments + rebasing on top of Chromium 99-based Brave (`master`). PTAL when you have a moment, thanks!

**Comment by mariospr** [comment]
@bridiver @iefremov @mkarolin Sorry to insist, but if you could review this that would be awesome. It should be passing on all platforms now (except Mac, which don't run upstream tests for now as it was way too problematic) modulus some weird & unrelated to this PR problems I've seeing lately, and the comment by @bridiver in https://github.com/brave/brave-core/pull/11814#discussion_r807159439 has already been addressed.

Thanks!

**Review by mkarolin — APPROVED** [review]
LGTM 👍

**Comment by mariospr** [comment]
@bridiver @iefremov Failures seem unrelated to the PR, please take a look to this if you have a moment, let's see if we can eventually merge it (it's been a while already :innocent: )

**Review by bridiver — APPROVED** [review]

**Comment by mariospr** [comment]
> Would be nice to get this in.

Yes, I'm planning on getting this one landed ASAP as in "today or tomorrow". The latest build seems to have succeeded and the only failure is unrelated to upstream tests, so I'll rebase once again and will merge if it's all good again.

Sorry for the delay, we had to adapt this twice from cr98 to cr99 and now from cr99 to cr100, but should definitely land now.

**Comment by mariospr** [comment]
Finally merged!

**Pull request merged**
